File: src/migration/executor.ts
/**
 * Migration Executor
 * 
 * Executes migration plans with proper backup, file operations, and rollback support.
 */

import { promises as fs } from 'node:fs';
import path from 'node:path';
import * as p from '@clack/prompts';
import { toolIdSchema, trackedFileSchema } from '../store/schema.js';
import type { FileRecord } from '../types.js';
import { writeToCanonical } from './canonical-writer.js';
import type {
  MergeConflict,
  MigrationAction,
  MigrationContext,
  MigrationPlan,
  MigrationResult,
  MigrationStats,
  ParsedSetup,
} from './types.js';

type ManifestFile = Pick<FileRecord, 'path' | 'hash' | 'source'>;

interface ManifestShape {
  version: string;
  setupScope: 'project' | 'workspace' | 'global';
  setupType: 'project' | 'workspace';
  tools: string[];
  projectName: string;
  installedAt: string;
  files: ManifestFile[];
}

function isManifestFile(value: unknown): value is ManifestFile {
  const parsed = trackedFileSchema.safeParse(value);
  return parsed.success;
}

function normalizeManifest(raw: unknown): ManifestShape {
  const now = new Date().toISOString();
  const fallback: ManifestShape = {
    version: '0.1.0',
    setupScope: 'project',
    setupType: 'project',
    tools: [],
    projectName: 'Migrated Project',
    installedAt: now,
    files: [],
  };

  if (!raw || typeof raw !== 'object') return fallback;
  const data = raw as Record<string, unknown>;

  const setupScope = data.setupScope;
  const normalizedScope: ManifestShape['setupScope'] =
    setupScope === 'global' || setupScope === 'workspace' || setupScope === 'project'
      ? setupScope
      : 'project';

  const setupType = data.setupType;
  const normalizedSetupType: ManifestShape['setupType'] =
    setupType === 'workspace' || setupType === 'project'
      ? setupType
      : normalizedScope === 'workspace'
        ? 'workspace'
        : 'project';

  const tools = Array.isArray(data.tools)
    ? data.tools.filter((tool): tool is string => typeof tool === 'string')
    : [];

  const files = Array.isArray(data.files)
    ? data.files.filter(isManifestFile).map((file) => ({ path: file.path, hash: file.hash, source: file.source }))
    : [];

  return {
    version: typeof data.version === 'string' ? data.version : fallback.version,
    setupScope: normalizedScope,
    setupType: normalizedSetupType,
    tools,
    projectName: typeof data.projectName === 'string' ? data.projectName : fallback.projectName,
    installedAt: typeof data.installedAt === 'string' ? data.installedAt : fallback.installedAt,
    files,
  };
}

function toManifestFile(record: FileRecord): ManifestFile {
  return {
    path: record.path,
    hash: record.hash,
    source: record.source,
  };
}

export async function executeMigrationPlan(
  context: MigrationContext,
  plan: MigrationPlan
): Promise<MigrationResult> {
  const executedActions: MigrationAction[] = [];
  const errors: string[] = [];
  const warnings: string[] = [];
  const stats: MigrationStats = {
    filesCreated: 0,
    filesModified: 0,
    filesBackedUp: 0,
    filesSkipped: 0,
    conflictsResolved: 0,
    conflictsUnresolved: 0,
  };

  let backupPath: string | undefined;

  try {
    if (context.options.interactive) {
      await resolvePlanConflictsInteractively(plan);
    }

    // Step 1: Create backup directory
    if (!context.options.skipBackup) {
      backupPath = await createBackupDir(context.targetPath);
    }

    // Step 2: Execute backup actions first
    const backupActions = plan.actions.filter(a => a.type === 'backup');
    for (const action of backupActions) {
      try {
        await executeBackupAction(action, context, backupPath!);
        executedActions.push(action);
        stats.filesBackedUp++;
      } catch (error) {
        errors.push(`Failed to backup ${action.sourcePath}: ${error}`);
      }
    }

    // Step 3: Execute create actions
    const createActions = plan.actions.filter(a => a.type === 'create');
    for (const action of createActions) {
      try {
        await executeCreateAction(action, context);
        executedActions.push(action);
        stats.filesCreated++;
      } catch (error) {
        errors.push(`Failed to create ${action.targetPath}: ${error}`);
      }
    }

    // Step 4: Execute modify actions
    const modifyActions = plan.actions.filter(a => a.type === 'modify');
    for (const action of modifyActions) {
      try {
        await executeModifyAction(action, context);
        executedActions.push(action);
        stats.filesModified++;
      } catch (error) {
        errors.push(`Failed to modify ${action.targetPath}: ${error}`);
      }
    }

    // Step 5: Count conflicts
    stats.conflictsResolved = plan.conflicts.filter(c => c.resolved).length;
    stats.conflictsUnresolved = plan.conflicts.filter(c => !c.resolved).length;
    plan.canProceed = stats.conflictsUnresolved === 0;

    // Step 6: Create .ai-setup.json if it doesn't exist
    await createAiSetupManifest(context, plan, executedActions);

    return buildMigrationResult({
      success: errors.length === 0,
      plan,
      executedActions,
      ...(backupPath ? { backupPath } : {}),
      errors,
      warnings,
      stats,
    });

  } catch (error) {
    // Fatal error - try to rollback if possible
    errors.push(`Fatal error during migration: ${error}`);
    
    return buildMigrationResult({
      success: false,
      plan,
      executedActions,
      ...(backupPath ? { backupPath } : {}),
      errors,
      warnings,
      stats,
    });
  }
}

export async function executeMigrationToCanonical(options: {
  context: MigrationContext;
  plan: MigrationPlan;
  parsedSetups: ParsedSetup[];
}): Promise<MigrationResult> {
  const { context, plan, parsedSetups } = options;
  const executedActions: MigrationAction[] = [];
  const errors: string[] = [];
  const warnings: string[] = [];
  const stats: MigrationStats = {
    filesCreated: 0,
    filesModified: 0,
    filesBackedUp: 0,
    filesSkipped: 0,
    conflictsResolved: 0,
    conflictsUnresolved: 0,
  };

  const fileRecords: FileRecord[] = [];
  let backupPath: string | undefined;

  try {
    if (!context.options.skipBackup) {
      backupPath = await createBackupDir(context.targetPath);
    }

    if (backupPath) {
      for (const parsed of parsedSetups) {
        for (const file of parsed.files) {
          const backupAction: MigrationAction = {
            type: 'backup',
            sourcePath: file.path,
            targetPath: path.join('.ai-setup-backup', file.path),
            description: `Backup ${file.path}`,
            reason: 'Before canonical migration write',
          };

          try {
            await executeBackupAction(backupAction, context, backupPath);
            executedActions.push(backupAction);
            stats.filesBackedUp += 1;
          } catch {
            // Some parsers include virtual files; skip silently.
          }
        }
      }
    }

    for (const parsed of parsedSetups) {
      const canonicalResult = await writeToCanonical({
        targetDir: context.targetPath,
        parsedSetup: parsed,
        fileRecords,
        dryRun: context.options.preview === true,
      });

      const createdPaths = [
        ...canonicalResult.agents,
        ...canonicalResult.skills,
        ...canonicalResult.prompts,
        ...canonicalResult.rules,
        ...(canonicalResult.rootConfig ? [canonicalResult.rootConfig] : []),
      ];

      for (const targetPath of createdPaths) {
        executedActions.push({
          type: 'create',
          targetPath,
          description: `Create ${targetPath}`,
          reason: `Canonical migration from ${parsed.metadata.adapter || 'adapter'}`,
        });
      }

      for (const targetPath of canonicalResult.skipped) {
        executedActions.push({
          type: 'skip',
          targetPath,
          description: `Skip ${targetPath}`,
          reason: 'File already exists',
        });
      }

      stats.filesCreated += createdPaths.length;
      stats.filesSkipped += canonicalResult.skipped.length;
    }

    await createAiSetupManifest(context, plan, executedActions, fileRecords);

    return buildMigrationResult({
      success: errors.length === 0,
      plan,
      executedActions,
      ...(backupPath ? { backupPath } : {}),
      errors,
      warnings,
      stats,
    });
  } catch (error) {
    errors.push(`Fatal error during canonical migration: ${error}`);

    return buildMigrationResult({
      success: false,
      plan,
      executedActions,
      ...(backupPath ? { backupPath } : {}),
      errors,
      warnings,
      stats,
    });
  }
}

interface BuildMigrationResultInput {
  success: boolean;
  plan: MigrationPlan;
  executedActions: MigrationAction[];
  backupPath?: string;
  errors: string[];
  warnings: string[];
  stats: MigrationStats;
}

function buildMigrationResult(input: BuildMigrationResultInput): MigrationResult {
  return {
    success: input.success,
    plan: input.plan,
    executedActions: input.executedActions,
    ...(input.backupPath ? { backupPath: input.backupPath } : {}),
    errors: input.errors,
    warnings: input.warnings,
    stats: input.stats,
  };
}

type InteractiveResolution = 'keep-existing' | 'use-new' | 'manual-edit';

export async function resolvePlanConflictsInteractively(plan: MigrationPlan): Promise<void> {
  for (const conflict of plan.conflicts) {
    if (conflict.resolved) {
      continue;
    }

    p.note(
      renderConflictDiff(conflict),
      `Conflict in ${conflict.file || 'unknown file'}:${conflict.lineStart}-${conflict.lineEnd}`,
    );

    const resolution = await p.select({
      message: `How should ${conflict.file || 'this file'} be resolved?`,
      options: [
        { value: 'keep-existing', label: 'Keep Existing', hint: 'Preserve the current file contents' },
        { value: 'use-new', label: 'Use New', hint: 'Apply the incoming ai-setup version' },
        { value: 'manual-edit', label: 'Manual Edit', hint: 'Enter the merged content yourself' },
      ],
    });

    if (p.isCancel(resolution)) {
      throw new Error('Migration cancelled during interactive conflict resolution');
    }

    await applyConflictResolution(conflict, resolution as InteractiveResolution);
  }
}

async function applyConflictResolution(
  conflict: MergeConflict,
  resolution: InteractiveResolution,
): Promise<void> {
  switch (resolution) {
    case 'keep-existing':
      conflict.resolved = true;
      conflict.resolution = resolution;
      conflict.resolvedContent = getConflictSide(conflict, 'ours');
      return;
    case 'use-new':
      conflict.resolved = true;
      conflict.resolution = resolution;
      conflict.resolvedContent = getConflictSide(conflict, 'theirs');
      return;
    case 'manual-edit':
      await resolveManualEdit(conflict);
      return;
  }
}

async function resolveManualEdit(conflict: MergeConflict): Promise<void> {
  const edited = await p.text({
    message: `Enter merged content for ${conflict.file || 'the conflict'}`,
    initialValue: buildManualEditSeed(conflict),
    placeholder: 'Type the final merged content',
  });

  if (p.isCancel(edited)) {
    throw new Error('Migration cancelled during interactive conflict resolution');
  }

  conflict.resolved = true;
  conflict.resolution = 'manual-edit';
  conflict.resolvedContent = String(edited);
}

function buildManualEditSeed(conflict: MergeConflict): string {
  const existing = getConflictSide(conflict, 'ours');
  const incoming = getConflictSide(conflict, 'theirs');

  return [
    '<<<<<<< EXISTING',
    existing,
    '=======',
    incoming,
    '>>>>>>> NEW',
  ].join('\n');
}

function getConflictSide(
  conflict: MergeConflict,
  side: 'base' | 'ours' | 'theirs',
): string {
  const legacyValue = (conflict as MergeConflict & Record<string, unknown>)[side];

  if (side === 'base') {
    return conflict.baseContent || normalizeLegacyConflictSide(legacyValue);
  }

  if (side === 'ours') {
    return conflict.oursContent || normalizeLegacyConflictSide(legacyValue);
  }

  return conflict.theirsContent || normalizeLegacyConflictSide(legacyValue);
}

function normalizeLegacyConflictSide(value: unknown): string {
  if (Array.isArray(value)) {
    return value.join('\n');
  }

  return typeof value === 'string' ? value : '';
}

export function renderConflictDiff(conflict: MergeConflict): string {
  return [
    '--- Existing',
    getConflictSide(conflict, 'ours') || '(empty)',
    '+++ New',
    getConflictSide(conflict, 'theirs') || '(empty)',
  ].join('\n');
}

async function createBackupDir(targetPath: string): Promise<string> {
  const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
  const backupPath = path.join(targetPath, '.ai-setup-backup', `migration-${timestamp}`);
  
  await fs.mkdir(backupPath, { recursive: true });
  
  return backupPath;
}

async function executeBackupAction(
  action: MigrationAction,
  context: MigrationContext,
  backupPath: string
): Promise<void> {
  if (!action.sourcePath) return;
  
  const sourceFullPath = path.join(context.sourcePath, action.sourcePath);
  const backupFullPath = path.join(backupPath, action.targetPath);
  
  // Ensure backup directory exists
  await fs.mkdir(path.dirname(backupFullPath), { recursive: true });
  
  // Copy file to backup
  await fs.copyFile(sourceFullPath, backupFullPath);
}

async function executeCreateAction(
  action: MigrationAction,
  context: MigrationContext
): Promise<void> {
  const targetFullPath = path.join(context.targetPath, action.targetPath);
  
  // Ensure directory exists
  await fs.mkdir(path.dirname(targetFullPath), { recursive: true });
  
  // File creation is handled by parsers - this just ensures directory exists
  // The actual content is written during the merge phase
}

async function executeModifyAction(
  action: MigrationAction,
  context: MigrationContext
): Promise<void> {
  // Modifications are handled by parsers during merge
  // This ensures the directory structure exists
  const targetFullPath = path.join(context.targetPath, action.targetPath);
  await fs.mkdir(path.dirname(targetFullPath), { recursive: true });
}

async function createAiSetupManifest(
  context: MigrationContext,
  plan: MigrationPlan,
  executedActions: MigrationAction[],
  fileRecords?: FileRecord[]
): Promise<void> {
  const manifestPath = path.join(context.targetPath, '.ai-setup.json');
  
  let existingManifest = normalizeManifest(null);
  
  // Try to read existing manifest
  try {
    const content = await fs.readFile(manifestPath, 'utf-8');
    existingManifest = normalizeManifest(JSON.parse(content));
  } catch {
    // No existing manifest
  }
  
  // Update tools
  for (const adapter of plan.adapters) {
    if (!toolIdSchema.safeParse(adapter).success) {
      continue;
    }

    if (!existingManifest.tools.includes(adapter)) {
      existingManifest.tools.push(adapter);
    }
  }
  
  if (fileRecords && fileRecords.length > 0) {
    for (const record of fileRecords) {
      const existingFile = existingManifest.files.find((f) => f.path === record.path);
      if (existingFile) {
        existingFile.hash = record.hash;
        existingFile.source = record.source;
      } else {
        existingManifest.files.push(toManifestFile(record));
      }
    }
  } else {
    // Track migrated files
    for (const action of executedActions) {
      if (action.type === 'create' || action.type === 'modify') {
        const existingFile = existingManifest.files.find((f) => f.path === action.targetPath);
        if (!existingFile) {
          existingManifest.files.push({
            path: action.targetPath,
            hash: 'migrated', // Will be updated by doctor
            source: 'migration',
          });
        }
      }
    }
  }
  
  // Write manifest
  await fs.writeFile(manifestPath, JSON.stringify(existingManifest, null, 2));
}

export async function rollbackMigration(
  _context: MigrationContext,
  _backupPath: string
): Promise<void> {
  // Restore files from backup
  // This is a safety feature for failed migrations
  console.warn('Rolling back migration...');
  
  // Implementation would restore backed up files
  // For now, we just log the intent
}

File: packages/cli/internal/migration/executor.go
package migration

import (
	"fmt"
	"os"
	"path/filepath"
	"time"

	"github.com/rluisb/lazyai/packages/cli/internal/files"
	"github.com/rluisb/lazyai/packages/cli/internal/manifest"
	"github.com/rluisb/lazyai/packages/cli/internal/types"
)

// Execute runs the migration plan, creating files and backups as needed.
// Ported from executeMigrationPlan in src/migration/executor.ts.
func Execute(plan *MigrationPlan, ctx *MigrationContext) (*MigrationResult, error) {
	var executedActions []MigrationAction
	var errors []string
	var warnings []string
	stats := MigrationStats{}

	var backupPath string
	var err error

	// Step 1: Create backup directory.
	if !ctx.Options.SkipBackup && !ctx.Options.Preview {
		backupPath, err = createBackupDir(ctx.TargetPath)
		if err != nil {
			return nil, fmt.Errorf("create backup dir: %w", err)
		}
	}

	// Step 2: Execute actions in order.
	for _, action := range plan.Actions {
		switch action.Type {
		case ActionTypeBackup:
			if err := executeBackupAction(action, ctx, backupPath); err != nil {
				errors = append(errors, fmt.Sprintf("Failed to backup %s: %s", action.SourcePath, err))
				continue
			}
			executedActions = append(executedActions, action)
			stats.FilesBackedUp++

		case ActionTypeCreate:
			if err := executeCreateAction(action, ctx); err != nil {
				errors = append(errors, fmt.Sprintf("Failed to create %s: %s", action.TargetPath, err))
				continue
			}
			executedActions = append(executedActions, action)
			stats.FilesCreated++

		case ActionTypeModify:
			if err := executeModifyAction(action, ctx); err != nil {
				errors = append(errors, fmt.Sprintf("Failed to modify %s: %s", action.TargetPath, err))
				continue
			}
			executedActions = append(executedActions, action)
			stats.FilesModified++

		case ActionTypeSkip:
			executedActions = append(executedActions, action)
			stats.FilesSkipped++
		}
	}

	// Step 3: Count conflicts.
	for _, c := range plan.Conflicts {
		if c.Resolved {
			stats.ConflictsResolved++
		} else {
			stats.ConflictsUnresolved++
		}
	}

	// Step 4: Update the manifest.
	if err := updateManifest(ctx, plan, nil); err != nil {
		warnings = append(warnings, fmt.Sprintf("Failed to update manifest: %s", err))
	}

	return &MigrationResult{
		Success:         len(errors) == 0,
		Plan:            plan,
		ExecutedActions: executedActions,
		BackupPath:      backupPath,
		Errors:          errors,
		Warnings:        warnings,
		Stats:           stats,
	}, nil
}

// ExecuteToCanonical writes parsed data into the .ai/ canonical format.
// Ported from executeMigrationToCanonical in src/migration/executor.ts.
func ExecuteToCanonical(ctx *MigrationContext, plan *MigrationPlan, parsedSetups []ParsedSetup) (*MigrationResult, error) {
	var executedActions []MigrationAction
	var errors []string
	var warnings []string
	stats := MigrationStats{}
	var fileRecords []types.TrackedFile

	var backupPath string
	var err error

	if !ctx.Options.SkipBackup && !ctx.Options.Preview {
		backupPath, err = createBackupDir(ctx.TargetPath)
		if err != nil {
			return nil, fmt.Errorf("create backup dir: %w", err)
		}
	}

	// Backup source files.
	if backupPath != "" {
		for _, parsed := range parsedSetups {
			for _, file := range parsed.Files {
				if err := executeBackupAction(MigrationAction{
					Type:       ActionTypeBackup,
					SourcePath: file.Path,
					TargetPath: filepath.Join(".ai-setup-backup", file.Path),
				}, ctx, backupPath); err != nil {
					// Skip silently for virtual files.
					continue
				}
				stats.FilesBackedUp++
			}
		}
	}

	// Write canonical files for each parsed setup.
	for _, parsed := range parsedSetups {
		result, err := WriteCanonical(ctx.TargetPath, &parsed, &fileRecords, ctx.Options.Preview)
		if err != nil {
			errors = append(errors, fmt.Sprintf("Canonical write failed: %s", err))
			continue
		}

		for _, p := range result.Agents {
			executedActions = append(executedActions, MigrationAction{
				Type:        ActionTypeCreate,
				TargetPath:  p,
				Description: fmt.Sprintf("Create %s", p),
				Reason:      fmt.Sprintf("Canonical migration from %s", parsed.Metadata["adapter"]),
			})
		}
		for _, p := range result.Skills {
			executedActions = append(executedActions, MigrationAction{
				Type:        ActionTypeCreate,
				TargetPath:  p,
				Description: fmt.Sprintf("Create %s", p),
				Reason:      fmt.Sprintf("Canonical migration from %s", parsed.Metadata["adapter"]),
			})
		}
		for _, p := range result.Prompts {
			executedActions = append(executedActions, MigrationAction{
				Type:        ActionTypeCreate,
				TargetPath:  p,
				Description: fmt.Sprintf("Create %s", p),
				Reason:      fmt.Sprintf("Canonical migration from %s", parsed.Metadata["adapter"]),
			})
		}
		for _, p := range result.Rules {
			executedActions = append(executedActions, MigrationAction{
				Type:        ActionTypeCreate,
				TargetPath:  p,
				Description: fmt.Sprintf("Create %s", p),
				Reason:      fmt.Sprintf("Canonical migration from %s", parsed.Metadata["adapter"]),
			})
		}
		if result.RootConfig != "" {
			executedActions = append(executedActions, MigrationAction{
				Type:        ActionTypeCreate,
				TargetPath:  result.RootConfig,
				Description: fmt.Sprintf("Create %s", result.RootConfig),
				Reason:      fmt.Sprintf("Canonical migration from %s", parsed.Metadata["adapter"]),
			})
		}

		stats.FilesCreated += len(result.Agents) + len(result.Skills) + len(result.Prompts) + len(result.Rules)
		if result.RootConfig != "" {
			stats.FilesCreated++
		}
		stats.FilesSkipped += len(result.Skipped)
	}

	if err := updateManifest(ctx, plan, fileRecords); err != nil {
		warnings = append(warnings, fmt.Sprintf("Failed to update manifest: %s", err))
	}

	return &MigrationResult{
		Success:         len(errors) == 0,
		Plan:            plan,
		ExecutedActions: executedActions,
		BackupPath:      backupPath,
		Errors:          errors,
		Warnings:        warnings,
		Stats:           stats,
	}, nil
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

func createBackupDir(targetDir string) (string, error) {
	timestamp := time.Now().UTC().Format("2006-01-02T15-04-05")
	backupPath := filepath.Join(targetDir, ".ai-setup-backup", fmt.Sprintf("migration-%s", timestamp))
	if err := files.EnsureDir(backupPath); err != nil {
		return "", err
	}
	return backupPath, nil
}

func executeBackupAction(action MigrationAction, ctx *MigrationContext, backupPath string) error {
	if action.SourcePath == "" {
		return nil
	}

	src := filepath.Join(ctx.SourcePath, action.SourcePath)
	if !files.FileExists(src) {
		return nil
	}

	dst := filepath.Join(backupPath, action.TargetPath)
	if err := files.EnsureDir(filepath.Dir(dst)); err != nil {
		return err
	}
	return files.CopyFile(src, dst)
}

func executeCreateAction(action MigrationAction, ctx *MigrationContext) error {
	targetFullPath := filepath.Join(ctx.TargetPath, action.TargetPath)
	return files.EnsureDir(filepath.Dir(targetFullPath))
}

func executeModifyAction(action MigrationAction, ctx *MigrationContext) error {
	targetFullPath := filepath.Join(ctx.TargetPath, action.TargetPath)
	return files.EnsureDir(filepath.Dir(targetFullPath))
}

// updateManifest creates or updates .ai-setup.json with migrated file records.
func updateManifest(ctx *MigrationContext, plan *MigrationPlan, fileRecords []types.TrackedFile) error {
	if ctx.Options.Preview {
		return nil
	}

	store, err := manifest.ReadManifestOptional(ctx.TargetPath)
	if err != nil {
		return err
	}
	if store == nil || store.Meta.SchemaVersion == 0 {
		defaults := types.DefaultStoreData()
		defaults.Config.SetupScope = types.SetupScopeProject
		defaults.Config.TargetDir = ctx.TargetPath
		defaults.Config.ProjectName = filepath.Base(ctx.TargetPath)
		store = &defaults
	}

	now := time.Now().UTC().Format(time.RFC3339)
	store.Meta.LastUpdatedAt = now
	if store.Config.TargetDir == "" {
		store.Config.TargetDir = ctx.TargetPath
	}
	if store.Config.ProjectName == "" {
		store.Config.ProjectName = filepath.Base(ctx.TargetPath)
	}

	existingTools := map[types.ToolId]bool{}
	for _, tool := range store.Config.Tools {
		existingTools[tool] = true
	}
	for _, adapter := range plan.Adapters {
		toolID, ok := adapterToToolID(adapter)
		if ok && !existingTools[toolID] {
			store.Config.Tools = append(store.Config.Tools, toolID)
			existingTools[toolID] = true
		}
	}

	existingFiles := map[string]int{}
	for i, tracked := range store.Files {
		existingFiles[tracked.Path] = i
	}
	for _, record := range fileRecords {
		record.Status = types.FileStatusInstalled
		record.LastCheckedAt = now
		if record.InstalledAt == "" {
			record.InstalledAt = now
		}
		if idx, ok := existingFiles[record.Path]; ok {
			store.Files[idx] = record
			continue
		}
		store.Files = append(store.Files, record)
		existingFiles[record.Path] = len(store.Files) - 1
	}

	store.Sync.LastSyncAt = now
	store.Sync.Dirty = false

	return manifest.WriteManifest(ctx.TargetPath, store)
}

func adapterToToolID(adapter string) (types.ToolId, bool) {
	switch adapter {
	case "opencode":
		return types.ToolIdOpenCode, true
	case "claude-code":
		return types.ToolIdClaudeCode, true
	case "copilot":
		return types.ToolIdCopilot, true
	default:
		return "", false
	}
}

// Rollback is a placeholder for restoring backed-up files.
func Rollback(_ *MigrationContext, _ string) error {
	fmt.Fprintln(os.Stderr, "Rolling back migration...")
	return nil
}
